src/python/pants/bin/goal_runner.py

```diff
@@ -250,6 +250,10 @@ def _execute_engine(self):
       return 1
 
     engine = RoundEngine()
+
+    sorted_goal_infos = engine.sort_goals(self._context, self._goals)
```

Comment:
This means we actually do this sorting twice, because execute is a public API which means I can't change it to pass in the `sorted_goal_infos` instead of `self._goals`. I guess I could memoize it in the engine... But this also feels like an API which maybe shouldn't actually be public, because people probably shouldn't be poking the engine directly from plugins?

Reply:
Are you referring to https://github.com/pantsbuild/pants/blob/8975e5a40bece5a04a1f69c029d30e62e2160beb/src/python/pants/engine/legacy_engine.py#L17-L24? If so, I don't see why it cannot take a list of sorted goals. Last recently changes to it were 594389cf9061fd71efdb11b7bc25db9c1b0db107 and c8c77b7bdb6275419e61845cd930f65585996c70 which are > 3 years away.

Reply:
It would be changing goals from being a `List<Goal>` to a `List<GoalInfo>` which feels like a non-trivial API change... (Particularly given `GoalInfo` is currently not even a public type)